test/e2e-test/rule-test/util/test-rule-functions.js

```diff
@@ -0,0 +1,145 @@
+    'gender', 'use', 'preferred', 'status'
+]);
+
+var response = function(status, message='', reqJson=null, resJson=null) {
+    return { valid: status, errorMessage: message, reqJson, resJson };
+};
+
+var fhirR4Validation =  function(reqJson, resJson){
+    var result = fhir.validate(resJson);
+    if (!result.valid)
+        return response(false, JSON.stringify(result, null, "\t"));
+    return response(true, '', reqJson, resJson);
+};
+
+var onePatient = function(reqJson, resJson){
+    var resources = fhir.evaluate(resJson, 'Bundle.entry.resource.resourceType');
+    var patientCount = testUtils.countOccurences(resources,'Patient');
+    if( patientCount !== 1)
+        return response(false, 'The bundle contains ' + patientCount + ' Patient resources');
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+var noDefaultGuid = function(reqJson, resJson){
+    var ids = fhir.evaluate(resJson, 'Bundle.entry.resource.id');
+    var defaultGuidCount = testUtils.countOccurences(ids, testUtils.defaultGuid);
+    if(defaultGuidCount >= 1)
+        return response(false, 'The bundle contains ' + defaultGuidCount + ' default Guid(s) ' + testUtils.defaultGuid);
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+var noSameGuid = function(reqJson, resJson){
+    var resources = fhir.evaluate(resJson, 'Bundle.entry.resource');
+    var ids = [];
+    for(var index in resources){
+        ids.push(resources[index].resourceType + '/' + resources[index].id);
+    }
+    var duplicates = testUtils.findDuplicates(ids);
+    if(duplicates.length !== 0)
+        return response(false, 'The bundle contains some duplicate Guids: ' + duplicates.toString());
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+const __revealObjectValues = (target, object, level) => {
+    if (level >= MAX_REVEAL_DEPTH) {
+        throw new Error('Reveal depth exceeds limit.');
+    }
+    if (_.isObject(object)) {
+        const keys = Object.keys(object).filter(key => !NON_COMPARE_PROPERTY.has(key));
+        return keys.every(key => __revealObjectValues(target, object[key], level + 1));
+    }
+    else {
+        const value = object.toString();
+        // specially treate datetime data type
+        const dateTimeRegex = /^[1-9]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])(\s+)?T?((20|21|22|23|[0-1]\d):[0-5]\d(:[0-5]\d)?)?.*$/;
+        if (dateTimeRegex.test(value)) {
+            return true;
+        }
+        return target.includes(value);
+    }
+};
+
+/**
+ * Every required property value appeared in conversion result should also appear in origin data.
+ * Some special data type should be treated separately (for example, "DataTime", since it has been processed by helper function).
+ * TODO: But there are many property values are extended from mapping table's explicit attribute, so there need a better way.
+ */
+const backwardValueReveal = (reqJson, resJson) => {
```

Comment:
Perhaps we could find out a better name for it? "backward" seems more close to versions and this scenario looks more like source and target.

Reply:
What about using `origin` instead of `backward` ?

Reply:
Sure. I think it will be better.